applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/formfinding_updated_reference_strategy.hpp

```diff
@@ -0,0 +1,1062 @@
+// KRATOS  ___|  |                   |                   |
+//       \___ \  __|  __| |   |  __| __| |   |  __| _` | |
+//             | |   |    |   | (    |   |   | |   (   | |
+//       _____/ \__|_|   \__,_|\___|\__|\__,_|_|  \__,_|_| MECHANICS
+//
+//  License:		 BSD License
+//					 license: structural_mechanics_application/license.txt
+//
+//  Main authors:    Long Chen
+//
+
+//#if !defined(KRATOS_FORMFINDING_UPDATED_REFERENCE_STRATEGY )
+//#define KRATOS_FORMFINDING_UPDATED_REFERENCE_STRATEGY
+#if !defined(KRATOS_FORMFINDING_UPDATED_REFERENCE_STRATEGY )
+#define  KRATOS_FORMFINDING_UPDATED_REFERENCE_STRATEGY
+// System includes
+
+// External includes 
+#include "boost/smart_ptr.hpp"
+
+
+// Project includes 
+#include "includes/define.h"
+#include "includes/model_part.h"
+#include "solving_strategies/strategies/solving_strategy.h"
+#include "solving_strategies/convergencecriterias/convergence_criteria.h"
+#include "utilities/openmp_utils.h"
+
+//default builder and solver
+#include "solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver.h"
+//#include "solving_strategies/builder_and_solvers/residualbased_elimination_builder_and_solver.h"
+
+namespace Kratos
+{
+
+    ///@name Kratos Globals 
+    ///@{
+
+
+    ///@}
+    ///@name Type Definitions 
+    ///@{
+
+    ///@}
+
+
+    ///@name  Enum's 
+    ///@{
+
+
+    ///@}
+    ///@name  Functions 
+    ///@{
+
+
+
+    ///@}
+    ///@name Kratos Classes 
+    ///@{
+
+    /// Short class definition.
+
+    /// Detail class definition.
+
+    //URL[Example of use html]{ extended_documentation/no_ex_of_use.html}
+
+    //URL[Example of use pdf]{ extended_documentation/no_ex_of_use.pdf}
+
+    //URL[Example of use doc]{ extended_documentation/no_ex_of_use.doc}
+
+    //URL[Example of use ps]{ extended_documentation/no_ex_of_use.ps}
+
+
+    //URL[Extended documentation html]{ extended_documentation/no_ext_doc.html}
+
+    //URL[Extended documentation pdf]{ extended_documentation/no_ext_doc.pdf}
+
+    //URL[Extended documentation doc]{ extended_documentation/no_ext_doc.doc}
+
+    //URL[Extended documentation ps]{ extended_documentation/no_ext_doc.ps}
+
+
+    template<class TSparseSpace,
+    class TDenseSpace, // = DenseSpace<double>,
+    class TLinearSolver //= LinearSolver<TSparseSpace,TDenseSpace>
+    >
+    class FormfindingUpdatedReferenceStrategy
```

Comment:
Hi Long Chen,
   how would it look like if you rewrote the main? (i mean, in pseudo code).

The reason for which i ask this, is because your problem may be sufficiently different to have a different "main"...

In any case what i am telling is that you should avoid copying the NR and just change a few lines. Instead you should see if it is possible to derive from it and only reimplement the few lines that are needed to change the behaviour.